### PR TITLE
Extend Redcution Expression and Fill Statement to Allow Inclusion of Center

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description

Some reductions in the ICON code that have the same start and end location include the center value in the reduction. This can not really be emulated with a trailing (or leading) additional operation if sparse dimensions are involved. That is, the entry for the central location in the sparse dimension would need to be extracted, involving manipulation of the memory of the sparse dimension, which we want to avoid. Hence, this PR:

* adds an option to include the center to the reduction expression
* adds an  option to include the center to the loop statement
* adds a field to sparse dimensions to indicate whether they are intended to be used in the reductions/loop statements including or excluding the center. This is type checked in the `UnstructuredDimensionsChecker`.

### Resolves / Enhances

Fixes issue #1000 

### Example

An example where this is needed can for example be found in `mo_math_gradients.f90` in the function `grad_green_gauss_cell_dycore`

### Testing

Quite a few tests have been added:

* dawn4py test
* integration tests using the `IIRBuilder` with output verification
* tests ensuring proper de/serialization of the newly introduced properties
* tests exercising the `UnstructuredDimensionsChecker` 

Unfortunately some things like the ico-cuda code generation have been tested manually only. The only way to inject this into the current CI would be `icondusk-e2e`, but dusk is currently not able to generate fields/reductions/loop statements including the center. 

### Dependencies

This PR is independent.

### CODE REVIEW

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Review by Stagno — COMMENTED** [review]
You are adding `includeCenter` in 3 classes:
- reduction
- sparse loop
- sparse field descriptor

and the corresponding protobuf, IIRBuilder methods, dawn4py bindings.
It feels like there's repetition that can be avoided.

I propose an alternative:
Replace all neighbor chains (`std::vector<LocationType>`) with an `UnstructuredIterationSpace`
```
struct UnstructuredIterationSpace {
	std::vector<LocationType> NeighborChain;
	bool IncludeCenter = false;

	UnstructuredIterationSpace(std::vector<LocationType> chain) : NeighborChain(chain) {}
	UnstructuredIterationSpace(std::vector<LocationType> chain, bool includeCenter) : NeighborChain(chain), IncludeCenter(includeCenter) { 
		if(includeCenter && chain.front() != chain.back()) 
			throw std::syntax_error("..."); 
	}
	operator std::vector<LocationType>() const { return NeighborChain; }
}
```

- encapsulates the concept of iteration space (neighborhood + whether to include center or not), instead of scattering it across its container (reduction, sparse loop, or sparse field descriptor) 
- also encapsulates `includeCenter && chain.front() != chain.back()` check instead of repeating it
- can directly compare `iterSpace1 == iterSpace2` (e.g. merging of sparse iterations in the future)
- cast-able (implicitly) into `std::vector<LocationType>` for those analyses that don't care if the center is included or not, and viceversa can cast implicitly a `std::vector<LocationType>` into `UnstructuredIterationSpace` (will be constructed with `IncludeCenter` false as default).
- dont' need to explictly set the "including center" context when visiting reduction/sparse loop, e.g. cuda-ico codegen's `parentIterationIncludesCenter*_`, the info would be retrievable from the current iteration space (substituting the current neighbor chain)

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Review by Stagno — CHANGES_REQUESTED** [review]
Few last remarks. But overall it's good.

- The serialized representation, in which you have the unstructured iteration space encapsulated in one message, is translated into constructor calls that take `chain, includeCenter` pairs. Is there a reason to separate the two again in such calls? (the alternative would be converting directly `protoIterSpace` into an `UnstructuredIterationSpace`) Has dawn4py anything to do with that?

- Also I noticed that toylib was not updated wrt this change.

File: dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp [hunk 1]
```diff
@@ -401,8 +425,8 @@ static void allocTempFields(MemberFunction& ctor, const iir::Stencil& stencil) {
       } else {
         ctor.addStatement("::dawn::allocField(&" + fname + "_, " + "mesh_." +
                           locToDenseSizeStringGpuMesh(hdims.getDenseLocationType()) + ", " +
-                          chainToSparseSizeString(hdims.getNeighborChain()) + ", " + kSizeStr +
-                          ")");
+                          chainToSparseSizeString(hdims.getIterSpace()) +
+                          (hdims.getIncludeCenter() ? "+1" : "") + ", " + kSizeStr + ")");
```
[the remark this answers is not in the document]
Comment (mroethlin) [reply dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp:429]: good catch!

**Review by Stagno — APPROVED** [review]
One optional improvement to consider, but it's good to go already.

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Pull request merged**